FitnessClub/Data/BLL/Services/CoachService.cs

```diff
@@ -0,0 +1,135 @@
+﻿using FitnessClub.Data.BLL.Interfaces;
+using FitnessClub.Data.DAL.Interfaces;
+using FitnessClub.Data.DAL.Utility;
+using FitnessClub.Data.Models.ViewModels;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace FitnessClub.Data.BLL.Services
+{
+    public class CoachService : EmployeeService, ICoachService
+    {
+        private readonly ISessionRepository sessionRepository;
+        private readonly ICoachRepository coachRepository;
+        private readonly int userId;
+        public CoachService(
+            ISessionRepository sessionRepository,
+            IEmployeeRepository employeeRepository,
+            ICoachRepository coachRepository,
+            UserResolverService userResolverService
+            ) : base(sessionRepository, employeeRepository, userResolverService)
+        {
+            this.sessionRepository = sessionRepository;
+            this.coachRepository = coachRepository;
+            userId = userResolverService.GetUserId();
+        }
+
+        public async Task AssignToSession(int coachId, int sessionId)
+        {
+            var coach = await coachRepository.GetById(coachId);
+            var session = sessionRepository.FindWithCoach(sessionId);
+            if (coach == null )
+            {
+                Serilog.Log.Information($"Couldn't find the coach with id {coachId}.");
+                return;
+            }
+            else if (session == null)
+            {
+                Serilog.Log.Information($"Couldn't find the session with id {sessionId}.");
+                return;
+            }
+            if (session.Coach != null)
+            {
+                Serilog.Log.Information($"Session already has assigned coach. Removed coach with id {session.Coach.PersonID} from the session and assigned the new one.");
+            }
+            session.Coach = coach;
+            session.CoachID = coach.PersonID;
+
+            await sessionRepository.Commit();
+        }
+
+        public async Task UnassignFromSession(int coachId, int sessionId)
+        {
+            var coach = coachRepository.FindWithSessions(coachId);
+            var session = coach.Sessions.Where(a => a.SessionID == sessionId).FirstOrDefault();
+            if (coach == null || session == null)
+            {
+                Serilog.Log.Information($"Couldn't find the session with given session id {sessionId} and person id {coachId}");
+                return;
+            }
+
+            var result = coach.Sessions.Remove(session);
+            if (!result)
+            {
+                coachRepository.Dispose();
```

Comment:
Why?

Reply:
Back when I was coding these classes, I wanted to ensure that if something goes wrong with the remove operation, the unit of work (and database context thats in it) would get disposed of, preventing any errors with next operations done on unit of work. 

Perhaps I mistunderstood when disposing is supposed to happen, and dependancy injection scoped lifetime already is handling the problem.